src/lib/device/virtual_dev.cpp

```diff
@@ -795,208 +447,65 @@ void VirtualDev::cp_flush() {
-            for (auto& chunk : chunk_list) {
-                if (chunk->blk_allocator_mutable()->verify_debug_bm(free_debug_bm) == false) {
-                    LOGERROR("Verify bitmap failure for chunk {}", static_cast< void* >(chunk));
-                    return BlkAllocStatus::FAILED;
-                } else {
-                    LOGDEBUG("Verify bitmap success for chunk {}", static_cast< void* >(chunk));
-                }
-            }
-        }
-        return BlkAllocStatus::SUCCESS;
-    } catch (const std::exception& e) {
-        LOGERROR("exception happened {}", e.what());
-        return BlkAllocStatus::FAILED;
-    }
+uint32_t VirtualDev::atomic_page_size() const {
+    auto* pdev = *(m_pdevs.begin());
+    return pdev->atomic_page_size();
 }
 
-/* Adds a primary chunk to the chunk list in pdev */
-void VirtualDev::add_primary_chunk(PhysicalDevChunk* chunk) {
-    auto const pdev_id = chunk->physical_dev()->dev_id();
-
-    if (m_chunk_size == 0) {
-        m_chunk_size = chunk->size();
-    } else {
-        HS_DBG_ASSERT_EQ(m_chunk_size, chunk->size());
-    }
-
-    if (!m_default_chunk || (m_default_chunk->chunk_id() > chunk->chunk_id())) { m_default_chunk = chunk; }
-
-    {
-        std::unique_lock< std::mutex > lk(m_free_streams_lk);
-        m_free_streams.push_back(chunk);
-    }
-    pdev_chunk_map* found_pcm = nullptr;
-    for (auto& pcm : m_primary_pdev_chunks_list) {
-        if (pcm.pdev->dev_id() == pdev_id) {
-            found_pcm = &pcm;
-            break;
-        }
-    }
-
-    if (found_pcm) {
-        found_pcm->chunks_in_pdev.push_back(chunk);
-    } else {
-        // Have not seen the pdev before, so add the chunk and also add it to device selector
-        pdev_chunk_map pcm;
-        pcm.pdev = m_mgr->get_pdev(pdev_id);
-        pcm.chunks_in_pdev.push_back(chunk);
+std::string VirtualDev::to_string() const { return ""; }
 
-        m_primary_pdev_chunks_list.push_back(pcm);
-        m_selector->add_pdev(pcm.pdev);
-    }
-
-    auto const max_chunk_size = MAX_DATA_CHUNK_SIZE(m_blk_size);
-    HS_DBG_ASSERT_LE(m_chunk_size, max_chunk_size);
-    std::shared_ptr< BlkAllocator > ba =
-        create_blk_allocator(m_allocator_type, block_size(), phys_page_size(), align_size(), m_chunk_size,
-                             m_auto_recovery, chunk->chunk_id(), m_recovery_init);
-    chunk->set_blk_allocator(ba);
-    if (!m_recovery_init) { chunk->recover(); }
-
-    /* set the same blk allocator to other mirror chunks */
-    auto const itr_pair = m_mirror_chunks.emplace(chunk, std::vector< PhysicalDevChunk* >{});
-    if (itr_pair.second) {
-        // Not found, just created a new entry
-    } else {
-        for (auto* chunk : itr_pair.first->second) {
-            chunk->set_blk_allocator(ba);
-        }
-    }
+shared< Chunk > VirtualDev::get_next_chunk(cshared< Chunk >& chunk) const {
+    return m_all_chunks[chunk->vdev_ordinal() % m_all_chunks.size()];
```

Comment:
```
    chunk->set_vdev_ordinal(m_all_chunks.size());
    m_pdevs.insert(chunk->physical_dev_mutable());
    m_all_chunks.push_back(chunk);
   ```
chunk0<vdev_ordinal starts from zero,  I believe we need
 `( chunk->vdev_ordinal() +1 ) % m_all_chunks.size()`

Reply:
we dont need lock on read path because we believe no r/w conflict on `m_all_chunks`?

Reply:
Yes on ordinal part (it needs a +1). Good catch.

On the r/w conflict m_all_chunks, as part of flag if dynamically adding chunk support (needed for journal vdev). If so we will use rcu to protect this.